src/daft-csv/src/local.rs

```diff
@@ -0,0 +1,924 @@
+                        &parse_options,
+                        reader,
+                        projection_indices,
+                        fields,
+                        read_daft_fields,
+                        read_schema,
+                        &mut buffer,
+                        include_columns,
+                        predicate,
+                        limit,
+                    );
+                    tx.send(tables)
+                        .expect("OneShot Channel should still be open");
+                });
+                rx.await
+            })
+        })
+        .buffered(n_threads)
+        .map(|v| {
+            v.context(JoinSnafu {})?
+                .context(super::OneShotRecvSnafu {})?
+        });
+    let flattened = stream
+        .map(|result: DaftResult<Vec<Table>>| {
+            let tables = result?;
+            DaftResult::Ok(futures::stream::iter(tables.into_iter().map(Ok)))
+        })
+        .try_flatten();
+
+    Ok(flattened)
+}
+
+/// A helper struct that implements `std::io::Read` over a slice of ChunkStates.
+struct MultiSliceReader<'a> {
+    states: &'a [ChunkState],
+    curr_read_idx: usize,
+    curr_read_offset: usize,
+}
+
+impl<'a> MultiSliceReader<'a> {
+    fn new(states: &'a [ChunkState]) -> Self {
+        Self {
+            states,
+            curr_read_idx: 0,
+            curr_read_offset: 0,
+        }
+    }
+}
+
+impl<'a> Read for MultiSliceReader<'a> {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        let buf_len = buf.len();
+        let mut position = 0;
+        while self.curr_read_idx < self.states.len() && position < buf_len {
+            let state = &self.states[self.curr_read_idx];
+            let (start, end, guard) = match state {
+                ChunkState::Start { slab, start, end } => {
+                    let guard = slab.read();
+                    (*start, *end, guard)
+                }
+                ChunkState::Continue { slab, end } => {
+                    let guard = slab.read();
+                    (0, *end, guard)
+                }
+                ChunkState::Final { slab, end, .. } => {
+                    let guard = slab.read();
+                    (0, *end, guard)
+                }
+            };
+            let slice = &guard.buffer[start..end];
+            if self.curr_read_offset < slice.len() {
+                let read_size = (buf_len - position).min(slice.len() - self.curr_read_offset);
+                buf[position..position + read_size].copy_from_slice(
+                    &slice[self.curr_read_offset..self.curr_read_offset + read_size],
+                );
+                self.curr_read_offset += read_size;
+                position += read_size;
+            }
+            if self.curr_read_offset >= slice.len() {
+                self.curr_read_offset = 0;
+                self.curr_read_idx += 1;
+            }
+        }
+        Ok(position)
+    }
+}
+
+const NEWLINE: u8 = b'\n';
```

Comment:
Should these be user configurable?

Reply:
As discussed offline, Daft's `read_csv` API currently does not take in non-\n record terminators. I added a TODO as a comment as a reminder that we should add this option to the API and pass it down here.